src/graph/sampler.cc

```diff
@@ -182,8 +184,11 @@ void GetUniformSample(const dgl_id_t* edge_id_list,
 
 /*
  * Non-uniform sample via ArrayHeap
+ *
+ * \param probability Transition probability on the entire graph, indexed by edge ID
  */
-void GetNonUniformSample(const float* probability,
+template<typename ValueType>
+void GetNonUniformSample(const ValueType* probability,
```

Comment:
Still suggest use `NDArray` for some of these arguments (IdArray for vid_list etc.). This could save argument like `ver_len` and make it potentially cleaner. Also, could you change dgl_id_t to template argument as well? We might have int32 graph in the future. You could use ATEN_ID_TYPE_SWITCH macro to dispatch the corresponding type.

Reply:
Actually those are *views* of the `vid_list`.  You can see the usage in L433-L450: what the routines actually took in is something like `val_list[indptr[dst_id]:indptr[dst_id] + ver_len]`.

Reply:
Replied in the other issue. Basically, for scalar type argument, use int64 is safe. For ndarray, it is type agnostic and the pointer will only be casted in the template implementation.

Reply:
Then should we have two IdType's `VIDType` and `EIDType`?  Technically a graph with int32 node IDs can still have more than 2B edges (i.e. int64 edge IDs).  The same question is raised in #699 but somehow it is not addressed.

Reply:
We decided to simplify the situation. If a graph is 32bit, then both its number of nodes and edges will not exceed 2^32. Otherwise, a 64 bit graph is used. This means that all the ndarrays in a graph must have the same bit width, making it easier to maintain. So using one id type is fine.

Reply:
Oh OK I see what you meant.  Can we do it in another PR?  We need to template-ize layer sampler and NodeFlow construction routines as well.  I feel this is beyond the scope of this PR.

Reply:
Sure